tests/Unit/TestMain.cpp

```diff
@@ -0,0 +1,45 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <catch2/catch_session.hpp>
+#include <catch2/catch_tostring.hpp>
+#include <catch2/reporters/catch_reporter_event_listener.hpp>
+#include <catch2/reporters/catch_reporter_registrars.hpp>
+#include <cstddef>
+#include <memory>
+
+#include "Framework/SetupLocalPythonEnvironment.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+
+class TestRunListener : public Catch::EventListenerBase {
+ public:
+  using Catch::EventListenerBase::EventListenerBase;
+
+  void fatalErrorEncountered(Catch::StringRef error) { ERROR(error); }
+};
+
+CATCH_REGISTER_LISTENER(TestRunListener)
+
+// Charm looks for this function but since we build without a main function or
+// main module we just have it be empty
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-declarations"
+extern "C" void CkRegisterMainModule(void) {}
+#pragma GCC diagnostic pop
+
+int main(int argc, char* argv[]) {
+  Catch::StringMaker<double>::precision =
+      std::numeric_limits<double>::max_digits10;
+  Catch::StringMaker<float>::precision =
+      std::numeric_limits<float>::max_digits10;
+
+  const int result = Catch::Session().run(argc, argv);
+
+  // In the case where we run all the non-failure tests at once we must ensure
```

Comment:
[optional] I think this comment can get updated as we remove more of the old-style error tests and use `CHECK_THROWS_WITH` instead. Probably something like:
```cpp
// Clean up the python environment only after all tests have finished running,
// since there could be multiple tests run in a single executable launch.
```

If you decide to do this, the comment is also in the `TestCharmMain.cpp` file